futures-util/src/future/abortable.rs

```diff
@@ -0,0 +1,131 @@
+    /// `AbortRegistration`s can be acquired through `AbortHandle::new`.
+    ///
+    /// When `abort` is called on the handle tied to `reg` or if `abort` has
+    /// already been called, the future will complete immediately without making
+    /// any further progress.
+    pub fn new(future: T, reg: AbortRegistration) -> Self {
+        Abortable {
+            future: future,
+            inner: reg.inner,
+        }
+    }
+
+    unsafe_pinned!(future: T);
+}
+
+/// A registration handle for a `Abortable` future.
+/// Values of this type can be acquired from `AbortHandle::new` and are used
+/// in calls to `Abortable::new`.
+#[derive(Debug, Clone)]
+pub struct AbortRegistration {
+    inner: Arc<AbortInner>,
+}
+
+/// A handle to a `Abortable` future.
+#[derive(Debug, Clone)]
+pub struct AbortHandle {
+    inner: Arc<AbortInner>,
+}
+
+impl AbortHandle {
+    /// Creates a new `AbortHandle` which can be used to abort a running future.
+    /// This function also returns a `AbortRegistration` which is used to
+    /// pair the handle with a `Abortable` future.
+    ///
+    /// This function is usually paired with a call to `Abortable::new`.
+    pub fn new() -> (Self, AbortRegistration) {
+        let inner = Arc::new(AbortInner {
+            waker: AtomicWaker::new(),
+            cancel: AtomicBool::new(false),
+        });
+
+        (
+            AbortHandle {
+                inner: inner.clone(),
+            },
+            AbortRegistration {
+                inner: inner,
+            },
+        )
+    }
+}
+
+// Inner type storing the waker to awaken and a bool indicating that it
+// should be cancelled.
+#[derive(Debug)]
+struct AbortInner {
+    waker: AtomicWaker,
+    cancel: AtomicBool,
+}
+
+/// Creates a new `Abortable` future and a `AbortHandle` which can be used to stop it.
+///
+/// This function is a convenient (but less flexible) alternative to calling
+/// `AbortHandle::new` and `Abortable::new` manually.
+pub fn abortable<T>(future: T) -> (Abortable<T>, AbortHandle)
+    where T: Future
+{
+    let (handle, reg) = AbortHandle::new();
+    (
+        Abortable::new(future, reg),
+        handle,
+    )
+}
+
+/// Indicator that the `Abortable` future was aborted.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Aborted;
+
+impl<T> Future for Abortable<T> where T: Future {
+    type Output = Result<T::Output, Aborted>;
+
+    fn poll(mut self: PinMut<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        // Check if the future has been aborted
+        if self.inner.cancel.load(Ordering::Relaxed) {
+            return Poll::Ready(Err(Aborted))
+        }
+
+        // attempt to complete the future
+        if let Poll::Ready(x) = self.future().poll(cx) {
+            return Poll::Ready(Ok(x))
+        }
+
+        // Register to receive a wakeup if the future is aborted in the... future
+        self.inner.waker.register(cx.waker());
```

Comment:
Does `AtomicWaker` support registering multiple wakers across different tasks? (I would assume not otherwise there would need to be some explicit de-registration?)

The fact that `AbortRegistration` is clonable means you can do something like

```rust
    let (_a_tx, a_rx) = oneshot::channel::<()>();
    let (_b_tx, b_rx) = oneshot::channel::<()>();
    let (abort_handle, reg) = AbortHandle::new();
    let a_rx = Abortable::new(a_rx, reg.clone());
    let b_rx = Abortable::new(b_rx, reg);
```

but it seems like that won't wake both when aborted.

Reply:
The `AtomicWaker::register` docs say:
> This function is safe to call concurrently, but this is generally a bad idea. Concurrent calls to `register` will attempt to register different tasks to be notified. One of the callers will win and have its task set, but there is no guarantee as to which caller will succeed.

Well spotted! I agree the scenario in your code sample won't work as intended. `abort` will always wake just one task of the two tasks, but it should wake both.

Reply:
Done.